### DISCUSSION TITLE
v0.10.0

### DISCUSSION THREAD

**Discussion by laipz8200** [langgenius/dify#9562] [Releases]
## 🎉 What's New in v0.10.0? 🎉

Hello to all the innovators out there! We're excited to unveil v0.10.0, a powerhouse update designed to enhance your flexibility and efficiency. Let’s delve into the highlights:

![features](https://github.com/user-attachments/assets/2f461dbb-a60a-433e-bb90-1c83ecd989cf)

### 🌟 Enhanced Multi-Modal Support

This central feature boosts your workflows and chatflows with greater flexibility:

- **Advanced File Handling**: Nodes like **START**, **LLM**, **HTTP Request** and **IF-ELSE** now seamlessly manage "File" types, simplifying the processing of various file formats within your workflows.
  
- **New Functional Nodes**:
  - **Document Extractor**: Converts documents into text effortlessly, making data extraction straightforward.
  - **List Operator**: Empowers you with operations to filter, sort, and limit arrays, enhancing data control and manipulation.

- **Efficient Array Handling**: Directly use `array[string]` and `array[number]` in LLM prompts for more agile data interactions.

Thanks to the tireless efforts of @zxhlyh in #9526 and @laipz8200 in #8818, this feature marks a significant leap forward.

Read more: [Introducing Dify Workflow File Upload: A Demo on AI Podcast](https://dify.ai/blog/introducing-dify-workflow-file-upload-a-demo-on-ai-podcast)

### ✉️ Registry and Login with Email

Simplifying user access with:

- **Email Registration & Login**: Offers a secure, straightforward access method for users, optimizing the onboarding process. Shoutout to @douxc in #8119 and @ZhouhaoJiang in #8120.

### 🚀 Additional Features

- **Dataset Recall Testing API**: Gain insights into data retrieval efficiency with this new API, brought to you by @gubinjie in #9300.
- **Fireworks Model Integration**: Access Qwen2.5 72B Instruct models for expanded AI capability, thanks to @yaoice in #9333 and #9340.

### 🛠️ Enhancements

- **Storybook Component Testing**: Streamlined UI testing and visualization, expertly structured by @xuzuodong in #9324.
- **Timezone Conversion Enhancements**: Enhanced precision and usability with support in the Time Tool, thanks to @hwzhuhao in #9393.

### 🔧 Fixes

- **Azure OpenAI Fixes**: Resolved key issues with max token calculations for improved performance, managed by @somethingwentwell in #9326.
- **Endpoint Functionality Fixes**: Strengthened the stability of the db-pool-stat endpoint, addressed by @hwzhuhao in #9478.

v0.10.0 brings a wealth of innovations to simplify and elevate your workflow capabilities. Dive in and explore these enhancements to harness the full potential of AI-driven tasks. Your feedback is pivotal, so keep it coming. Happy building! 🚀

---

### Docker compose deployments

1. Back up your customized docker-compose YAML file (optional)

   ```bash
   cd docker
   cp docker-compose.yaml docker-compose.yaml.$(date +%s).bak
   ```

2. Get the latest code from the main branch

   ```bash
   git checkout main
   git pull origin main
   ```

3. Stop the service，Command, please execute in the docker directory

   ```bash
   docker compose down
   ```

4. Back up data

   ```bash
   tar -cvf volumes-$(date +%s).tgz volumes
   ```

5. Upgrade services

   ```bash
   docker compose up -d
   ```

### Source Code deployments

1. Stop API server, Worker and Web frontend Server.

2. Get the latest code from the release branch:

   ```bash
   git checkout 0.10.0
   ```

3. Update Python dependencies:

   ```bash
   cd api
   poetry install
   ```

7. Then, let's run the migration script:

   ```bash
   poetry run flask db upgrade
   ```

8. Finally, run API server, Worker and Web frontend Server again.

---

## What's Changed
* feat: added dataset recall testing API by @gubinjie in https://github.com/langgenius/dify/pull/9300
* feat: disable chat service API passing `parent_message_id` by @xuzuodong in https://github.com/langgenius/dify/pull/8984
* fix: poetry installation in CI jobs by @bowenliang123 in https://github.com/langgenius/dify/pull/9336
* feat: Add fireworks custom llm intergration by @yaoice in https://github.com/langgenius/dify/pull/9333
* feat: Add qwen2.5 72B Instruct model in Fireworks AI by @yaoice in https://github.com/langgenius/dify/pull/9340
* chore: clean-up unnecessary annotation on configs with non-null default value by @bowenliang123 in https://github.com/langgenius/dify/pull/9323
* refactor: remove unnecessary 'closing' usage for boto3 client by @y-omr in https://github.com/langgenius/dify/pull/9343
* Adapt input type parameter with MiniMax embedding model by @yaoice in https://github.com/langgenius/dify/pull/9342
* refactor: Add an enumeration type and use the factory pattern to obtain the corresponding class by @hwzhuhao in https://github.com/langgenius/dify/pull/9356
* fix: In the output, the order of 'ta' is sometimes reversed as 'at'. #8015 by @wsbao in https://github.com/langgenius/dify/pull/8791
* fix: Azure OpenAI o1 max_completion_token and get_num_token_from_messages error by @somethingwentwell in https://github.com/langgenius/dify/pull/9326
* feat: add glm-4-flashx, deprecated chatglm_turbo by @hjlarry in https://github.com/langgenius/dify/pull/9357
* fix: (#9336 followup) skip poetry preperation in style workflow when no change in api folder by @bowenliang123 in https://github.com/langgenius/dify/pull/9362
* feat: add minimax abab6.5t support by @cuckootan in https://github.com/langgenius/dify/pull/9365
* fix: add missing vikingdb param in docker .env.example by @yaoice in https://github.com/langgenius/dify/pull/9334
* fix: remove the stream option of zhipu and gemini by @hjlarry in https://github.com/langgenius/dify/pull/9319
* feat: Enable baiduvector intergration test by @yaoice in https://github.com/langgenius/dify/pull/9369
* fix: use gpt-4o-mini for validating credentials by @crazywoola in https://github.com/langgenius/dify/pull/9387
* feat: storybook by @xuzuodong in https://github.com/langgenius/dify/pull/9324
* fix: Deprecated gemma2-9b model in Fireworks AI Provider by @yaoice in https://github.com/langgenius/dify/pull/9373
* feat: add timezone conversion for time tool by @hwzhuhao in https://github.com/langgenius/dify/pull/9393
* Fix/economical knowledge retrieval by @JohnJyong in https://github.com/langgenius/dify/pull/9396
* fix: incorrect webapp image displayed by @yuyinws in https://github.com/langgenius/dify/pull/9401
* refactor: update the default values of top-k parameter in vdb to be consistent by @hwzhuhao in https://github.com/langgenius/dify/pull/9367
* fix: resolve overlap issue with API Extension selector and modal by @DDDDD12138 in https://github.com/langgenius/dify/pull/9407
* add clean 7 days datasets by @JohnJyong in https://github.com/langgenius/dify/pull/9424
* update dataset clean rule by @JohnJyong in https://github.com/langgenius/dify/pull/9426
* fix https://github.com/langgenius/dify/issues/9409 by @ybalbert001 in https://github.com/langgenius/dify/pull/9433
* Update dify-sandbox version to 0.2.10 by @hexian2001 in https://github.com/langgenius/dify/pull/9439
* Fix: rerank switch and validation before run by @YIXIAO0 in https://github.com/langgenius/dify/pull/9416
* revert yarnlock by @zxhlyh in https://github.com/langgenius/dify/pull/9445
* fix: enable Tencent Cloud Vector integration test by @yaoice in https://github.com/langgenius/dify/pull/9441
* Fix/firefox overflow error by @WTW0313 in https://github.com/langgenius/dify/pull/9438
* feat: add the workflow tool of comfyUI by @hjlarry in https://github.com/langgenius/dify/pull/9447
* fix: remove the undefined variable line by @hwzhuhao in https://github.com/langgenius/dify/pull/9446
* fix(workflow): Implement automatic variable addition from opening statement to start node by @Kevin9703 in https://github.com/langgenius/dify/pull/9450
* refactor: move the embedding to the rag module and abstract the rerank runner for extension by @hwzhuhao in https://github.com/langgenius/dify/pull/9423
* feat:add wenxin rerank by @chzphoenix in https://github.com/langgenius/dify/pull/9431
* feat: add yi-lightning llm model for yi by @hwzhuhao in https://github.com/langgenius/dify/pull/9458
* fix: add missing vector type to migrate command by @yaoice in https://github.com/langgenius/dify/pull/9470
* fix: update mismatch vector type by @yaoice in https://github.com/langgenius/dify/pull/9462
* feat: add parameter top-k for the llm model provided by openrouter and siliconflow by @hwzhuhao in https://github.com/langgenius/dify/pull/9455
* controller test by @s2oBCN in https://github.com/langgenius/dify/pull/9469
* fix fetch apps by @horochx in https://github.com/langgenius/dify/pull/9453
* fix: resolve the error with the db-pool-stat endpoint by @hwzhuhao in https://github.com/langgenius/dify/pull/9478
* feat: add yi custom llm intergration by @hwzhuhao in https://github.com/langgenius/dify/pull/9482
* Added Llama 3.2 Vision Models Speech2Text Models for Groq by @taowang1993 in https://github.com/langgenius/dify/pull/9479
* add yuqye（https://www.yuque.com）tools by @lichengwu in https://github.com/langgenius/dify/pull/8960
* feat: Add volcengine tos storage test by @yaoice in https://github.com/langgenius/dify/pull/9495
* chore: update the description for storage_type by @userzhangqg in https://github.com/langgenius/dify/pull/9492
* refine wording in license by @guchenhe in https://github.com/langgenius/dify/pull/9505
* fix: resolve the error of docker-compose startup when the storage is baidu-obs by @hwzhuhao in https://github.com/langgenius/dify/pull/9502
* Resolve 9508 openai compatible rerank by @ziyu4huang in https://github.com/langgenius/dify/pull/9511
* feat: Add custom username and avatar define in discord tool by @yaoice in https://github.com/langgenius/dify/pull/9514
* fix: ignore all files except for .gitkeep under docker/nginx/ssl by gitignore by @kurokobo in https://github.com/langgenius/dify/pull/9518
* refactor wenxin rerank by @chzphoenix in https://github.com/langgenius/dify/pull/9486
* fix: draft run workflow node with image will raise error by @hjlarry in https://github.com/langgenius/dify/pull/9406
* New Auth Methods by @douxc in https://github.com/langgenius/dify/pull/8119
* chore: translate i18n files by @github-actions in https://github.com/langgenius/dify/pull/9523
* Feat/new login by @ZhouhaoJiang in https://github.com/langgenius/dify/pull/8120
* Feat/attachments by @zxhlyh in https://github.com/langgenius/dify/pull/9526
* feat/enhance the multi-modal support by @laipz8200 in https://github.com/langgenius/dify/pull/8818
* fix(dockerfile): update libsqlite3 package version by @laipz8200 in https://github.com/langgenius/dify/pull/9530
* Fix: chatbot retrieval setting save issue by @YIXIAO0 in https://github.com/langgenius/dify/pull/9536
* fix: RESET_PASSWORD_TOKEN_EXPIRY_HOURS error by @ZhouhaoJiang in https://github.com/langgenius/dify/pull/9542
* fix: list filter node url not work by @hjlarry in https://github.com/langgenius/dify/pull/9540
* fix: list filter node some operator raise error by @hjlarry in https://github.com/langgenius/dify/pull/9539
* fix: input vars not auto rename in list operator by @iamjoel in https://github.com/langgenius/dify/pull/9550
* feat(file-upload): add support for optional file source parameter by @laipz8200 in https://github.com/langgenius/dify/pull/9554
* fix: TOKEN_EXPIRY_MINUTES by @ZhouhaoJiang in https://github.com/langgenius/dify/pull/9557
* fix: when upload file response 3** will raise error by @hjlarry in https://github.com/langgenius/dify/pull/9555
* chore: translate i18n files by @github-actions in https://github.com/langgenius/dify/pull/9528
* chore: update version to 0.10.0 in packaging and docker configurations by @laipz8200 in https://github.com/langgenius/dify/pull/9534

## New Contributors
* @wsbao made their first contribution in https://github.com/langgenius/dify/pull/8791
* @yuyinws made their first contribution in https://github.com/langgenius/dify/pull/9401
* @hexian2001 made their first contribution in https://github.com/langgenius/dify/pull/9439
* @chzphoenix made their first contribution in https://github.com/langgenius/dify/pull/9431
* @taowang1993 made their first contribution in https://github.com/langgenius/dify/pull/9479
* @userzhangqg made their first contribution in https://github.com/langgenius/dify/pull/9492

**Full Changelog**: https://github.com/langgenius/dify/compare/0.9.2...0.10.0

<hr /><em>This discussion was created from the release <a href='https://github.com/langgenius/dify/releases/tag/0.10.0'>v0.10.0</a>.</em>

**Comment by Smashhhhhhhhhh** [comment]
Although the 'Single File' is added to the input parameters of the workflow, why can't it be applied in the subsequent LLM? It's really a pity that there is no such function

**Reply by laipz8200** [reply]
We added a new `Doc Extractor` node to process document-typed files, other types will be supported later.

**Reply by Smashhhhhhhhhh** [reply]
> We added a new `Doc Extractor` node to process document-typed files, other types will be supported later.我们添加了一个新的`Doc Extractor`节点来处理文档类型的文件，稍后将支持其他类型。

It would be very useful if you can add it as soon as possible, thanks

**Reply by Smashhhhhhhhhh** [reply]
> We added a new `Doc Extractor` node to process document-typed files, other types will be supported later.

Multimodal LLM, all input parameters can be directly passed in url and mime/type, LLM will directly read and process them, I hope Dify can implement this function in 'LLM' Thank you

**Comment by jujuju102** [comment]
更新到0.10.0后，使用webapp图标替换应用界面的机器人图标功能失效了，更换webapp图标后，聊天界面机器人图标还是小机器人，之前版本是正常的。

**Reply by laipz8200** [reply]
Thanks for the report, we will fix this issue in the next version.

**Reply by jujuju102** [reply]
Thanks
